File: clients/android/app/src/main/java/local/visual/corepose/I18n.kt
package local.visual.corepose

import android.content.Context
import org.json.JSONObject

object I18n {
    const val DEFAULT = "en"
    val SUPPORTED = listOf("en", "es", "zh", "ko", "ja", "fr", "de", "it", "pt")
    val LABELS = mapOf(
        "en" to "English",
        "es" to "Español",
        "zh" to "中文",
        "ko" to "한국어",
        "ja" to "日本語",
        "fr" to "Français",
        "de" to "Deutsch",
        "it" to "Italiano",
        "pt" to "Português",
    )

    private const val PREFS = "visual_prefs"
    private const val KEY_LANG = "language"

    private var catalog: Map<String, Map<String, String>> = emptyMap()
    private var lang: String = DEFAULT
    private var appContext: Context? = null

    fun init(context: Context) {
        appContext = context.applicationContext
        lang = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY_LANG, DEFAULT)
            ?.takeIf { it in SUPPORTED }
            ?: DEFAULT
        catalog = loadCatalog(context)
    }

    fun language(): String = lang

    fun setLanguage(code: String) {
        val next = if (code in SUPPORTED) code else DEFAULT
        lang = next
        appContext?.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            ?.edit()
            ?.putString(KEY_LANG, next)
            ?.apply()
    }

    fun t(key: String, vars: Map<String, String> = emptyMap(), lang: String? = null): String {
        val code = lang ?: this.lang
        val raw = if (code == DEFAULT) {
            key
        } else {
            catalog[key]?.get(code) ?: key
        }
        return interpolate(raw, vars)
    }

    fun loadCatalogJson(json: String): Map<String, Map<String, String>> {
        val root = JSONObject(json)
        val strings = root.optJSONObject("strings") ?: return emptyMap()
        val out = HashMap<String, Map<String, String>>(strings.length())
        val keys = strings.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val node = strings.optJSONObject(key) ?: continue
            val langs = HashMap<String, String>()
            val langKeys = node.keys()
            while (langKeys.hasNext()) {
                val code = langKeys.next()
                langs[code] = node.optString(code)
            }
            out[key] = langs
        }
        return out
    }

    fun interpolate(template: String, vars: Map<String, String>): String {
        var out = template
        for ((name, value) in vars) {
            out = out.replace("{$name}", value)
            out = out.replace(Regex("\\{${Regex.escape(name)}:[^}]+\\}"), value)
        }
        return out
    }

    private fun loadCatalog(context: Context): Map<String, Map<String, String>> {
        return try {
            context.assets.open("strings.json").bufferedReader().use { loadCatalogJson(it.readText()) }
        } catch (_: Exception) {
            emptyMap()
        }
    }
}

File: clients/ios/VisualPose/I18n.swift
import Foundation

enum I18n {
    static let `default` = "en"
    static let supported = ["en", "es", "zh", "ko", "ja", "fr", "de", "it", "pt"]
    static let labels: [(String, String)] = [
        ("en", "English"),
        ("es", "Español"),
        ("zh", "中文"),
        ("ko", "한국어"),
        ("ja", "日本語"),
        ("fr", "Français"),
        ("de", "Deutsch"),
        ("it", "Italiano"),
        ("pt", "Português"),
    ]

    private static let langKey = "language"
    private static var catalog: [String: [String: String]] = [:]
    private static var lang: String = `default`

    static func language() -> String { lang }

    static func initCatalog(json: String, storedLang: String?) {
        catalog = loadCatalogJson(json)
        if let storedLang, supported.contains(storedLang) {
            lang = storedLang
        } else {
            lang = UserDefaults.standard.string(forKey: langKey).flatMap { supported.contains($0) ? $0 : nil } ?? `default`
        }
    }

    static func setLanguage(_ code: String) {
        lang = supported.contains(code) ? code : `default`
        UserDefaults.standard.set(lang, forKey: langKey)
    }

    static func t(_ key: String, vars: [String: String] = [:], lang code: String? = nil) -> String {
        let use = code ?? lang
        let raw: String
        if use == `default` {
            raw = key
        } else {
            raw = catalog[key]?[use] ?? key
        }
        return interpolate(raw, vars: vars)
    }

    static func loadCatalogJson(_ json: String) -> [String: [String: String]] {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let strings = root["strings"] as? [String: Any]
        else { return [:] }
        var out: [String: [String: String]] = [:]
        for (key, value) in strings {
            guard let node = value as? [String: Any] else { continue }
            var langs: [String: String] = [:]
            for (code, text) in node {
                if let text = text as? String { langs[code] = text }
            }
            out[key] = langs
        }
        return out
    }

    static func interpolate(_ template: String, vars: [String: String]) -> String {
        var out = template
        for (name, value) in vars {
            out = out.replacingOccurrences(of: "{\(name)}", with: value)
            if let regex = try? NSRegularExpression(pattern: "\\{\(NSRegularExpression.escapedPattern(for: name)):[^}]+\\}") {
                out = regex.stringByReplacingMatches(
                    in: out,
                    range: NSRange(out.startIndex..., in: out),
                    withTemplate: NSRegularExpression.escapedTemplate(for: value)
                )
            }
        }
        return out
    }
}
